pkg/auth/auth.go

```diff
@@ -0,0 +1,182 @@
+package auth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	imageAuth "github.com/containers/image/v5/pkg/docker/config"
+	"github.com/containers/image/v5/types"
+	dockerAPITypes "github.com/docker/docker/api/types"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// XRegistryAuthHeader is the key to the encoded registry authentication
+// configuration in an http-request header.
+const XRegistryAuthHeader = "X-Registry-Auth"
+
+// GetCredentials extracts one or more DockerAuthConfigs from the request's
+// header.  The header could specify a single-auth config in which case the
+// first return value is set.  In case of a multi-auth header, the contents are
+// stored in a temporary auth file (2nd return value).  Note that the auth file
+// should be removed after usage.
+func GetCredentials(r *http.Request) (*types.DockerAuthConfig, string, error) {
+	authHeader := r.Header.Get(XRegistryAuthHeader)
+	if len(authHeader) == 0 {
+		return nil, "", nil
+	}
+
+	// First look for a multi-auth header (i.e., a map).
+	authConfigs, err := multiAuthHeader(r)
+	if err == nil {
+		authfile, err := authConfigsToAuthFile(authConfigs)
+		return nil, authfile, err
+	}
+
+	// Fallback to looking for a single-auth header (i.e., one config).
+	authConfigs, err = singleAuthHeader(r)
+	if err != nil {
+		return nil, "", err
+	}
+	var conf *types.DockerAuthConfig
+	for k := range authConfigs {
+		c := authConfigs[k]
+		conf = &c
+		break
+	}
+	return conf, "", nil
+}
+
+// RemoveAuthfile is a convenience function that is meant to be called in a
+// deferred statement. If non-empty, it removes the specified authfile and log
+// errors.  It's meant to reduce boilerplate code at call sites of
+// `GetXRegistryAuth`.
+func RemoveAuthfile(authfile string) {
+	if authfile == "" {
+		return
+	}
+	if err := os.Remove(authfile); err != nil {
+		logrus.Errorf("Error removing temporary auth file %q: %v", authfile, err)
+	}
+}
+
+// EncodeSingleAuthConfig serializes the auth configuration as a base64 encoded JSON payload.
+func EncodeSingleAuthConfig(authConfig types.DockerAuthConfig) (string, error) {
+	conf := imageAuthToDockerAuth(authConfig)
+	buf, err := json.Marshal(conf)
+	if err != nil {
+		return "", err
+	}
+	return base64.URLEncoding.EncodeToString(buf), nil
+}
+
+// EncodeMultiAuthConfigs serializes the auth configurations as a base64 encoded JSON payload.
+func EncodeMultiAuthConfigs(authConfigs map[string]types.DockerAuthConfig) (string, error) {
```

Comment:
Usually this would be encoded as multiple headers, but I think that would break here because you would loose the registry name.  Maybe add a comment to that fact?

Reply:
That and it would break the projectatomic Docker and hence RHEL 7 -> 8 upgrade path.